### PR TITLE
check for side effects in quantifiers

### PR DESCRIPTION
**Pull request by kroening**
- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] My contribution is formatted in line with CODING_STANDARD.md.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by smowton — APPROVED** [review]

File: src/goto-programs/goto_clean_expr.cpp [hunk 1]
```diff
@@ -93,8 +93,6 @@ bool goto_convertt::needs_cleaning(const exprt &expr)
   // g1 = (i == 0)
   // g2 = (i > 10)
   // forall (i : int) (g1 || g2)
-  if(expr.id()==ID_forall || expr.id()==ID_exists)
-    return false;
```
Comment (tautschnig) [inline src/goto-programs/goto_clean_expr.cpp]: The comment above this code needs to be moved, it doesn't actually make sense when these lines are removed.

File: src/goto-programs/goto_clean_expr.cpp [hunk 2]
```diff
@@ -409,7 +407,7 @@ void goto_convertt::clean_expr(
     // check if there are side-effects
     goto_programt tmp;
     clean_expr(expr.op1(), tmp, mode, true);
-    if(tmp.instructions.empty())
+    if(!tmp.instructions.empty())
```
Comment (tautschnig) [inline src/goto-programs/goto_clean_expr.cpp]: So this was dead code before. It's no longer dead code now - do we have any tests exercising it?

**Review by tautschnig — CHANGES_REQUESTED** [review]

**Comment by kroening** [comment]
Now with different approach -- the rule is now checked in the front-end

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 0812cc17e3e057e2b6927a35bc18d186d814d3ef).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85673814

File: src/goto-programs/goto_clean_expr.cpp [hunk 3]
```diff
@@ -405,18 +405,7 @@ void goto_convertt::clean_expr(
...
-    clean_expr(expr.op1(), tmp, mode, true);
-    if(tmp.instructions.empty())
-    {
-      error().source_location=expr.find_source_location();
-      error() << "no side-effects in quantified expressions allowed"
-              << eom;
-      throw 0;
-    }
-    return;
+    // the front-end checks these for side-effects
```
Comment (smowton) [inline src/goto-programs/goto_clean_expr.cpp]: Invariant?

File: src/ansi-c/c_typecheck_expr.cpp [hunk 4]
```diff
@@ -166,6 +166,18 @@ bool c_typecheck_baset::gcc_types_compatible_p(
   return false;
 }
 
+static bool has_side_effect(const exprt &expr)
```
Comment (tautschnig) [inline src/ansi-c/c_typecheck_expr.cpp]: No need - see below.

File: src/ansi-c/c_typecheck_expr.cpp [hunk 5]
```diff
@@ -303,6 +315,13 @@ void c_typecheck_baset::typecheck_expr_main(exprt &expr)
       throw 0;
     }
 
+    if(has_side_effect(expr.op1()))
```
Comment (tautschnig) [inline src/ansi-c/c_typecheck_expr.cpp]: `if(has_subexpr(expr.op1(), ID_side_effect))`

File: src/goto-programs/goto_clean_expr.cpp (hunk 3 above)
Comment (tautschnig) [reply src/goto-programs/goto_clean_expr.cpp]: `DATA_INVARIANT(!has_subexpr(expr, ID_side_effect), "the front-end should check quantified expressions for side-effects");`

**Review by tautschnig — CHANGES_REQUESTED** [review]
I'd suggest to simplify as described below.

File: src/goto-programs/goto_clean_expr.cpp [hunk 6]
```diff
@@ -405,18 +406,7 @@ void goto_convertt::clean_expr(
...
-    clean_expr(expr.op1(), tmp, mode, true);
-    if(tmp.instructions.empty())
-    {
-      error().source_location=expr.find_source_location();
-      error() << "no side-effects in quantified expressions allowed"
-              << eom;
-      throw 0;
-    }
-    return;
+    DATA_INVARIANT(!has_subexpr(expr, ID_side_effect), "the front-end should check quantified expressions for side-effects");
```
Comment (tautschnig) [inline src/goto-programs/goto_clean_expr.cpp]: This line got too long.

**Review by tautschnig — APPROVED** [review]
Looks good, except for the line break.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: f505207b0a8b2bc50e55648876137bd674326cee).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85988064

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: a3fa91130b2b0d93cc1ccb363cf1c8947a479bcc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86019720

**Pull request merged**